TRT-1811: fallback queries

neisw opened the pull request:
- Queries previous releases without variant, capability or testid filtering
- Caches results so navigating down through capabilities doesn't cause the query to run again
- Compares stats from previous releases based on testIdentification and selects the one with the highest pass rate
- Includes Advanced Option for `Historical release analysis` default is exclude currently
- Fallback test details shows both the queried base release status and fallback when applicable
  - Override Release Assessment
  - Fallback Release Assessment
  - Fallback Release basis with dates
  - Override Release basis with dates (new tab)

![image](https://github.com/user-attachments/assets/d81f7c4e-0ec8-418a-90b1-f47a119ea455)

When included the regressed since column will be empty since regression tracking is currently not aware of these

![image](https://github.com/user-attachments/assets/7bed10aa-cf5f-4517-b9af-c37371c6c850)

Test Details Report will show tabs for the requested basis as well as the fallback / historical basis.  It will indicate an override as well as an explanation

![image](https://github.com/user-attachments/assets/0ac92f9b-33ae-4ea3-bdd9-eca2b867f477)

sosiouxme commented on sippy-ng/src/component_readiness/AdvancedOptions.js:
```diff
@@ -123,6 +129,13 @@ export default function AdvancedOptions(props) {
               name="ignoreDisruption"
               color="primary"
             />
+            <p>Fallback Basis: {ignoreFallbackBasis ? 'ignore' : 'keep'}</p>
+            <Switch
+              checked={ignoreFallbackBasis}
+              onChange={handleChangeIgnoreFallbackBasis}
+              name="ignoreFallbackBasis"
+              color="primary"
+            />
```
could it please have a tooltip to explain what it's for

sosiouxme replied:
also i would find this less confusing if the boolean were reversed and so we had "Enable Fallback Basis" defaulting to false (for now)

dgoodwin replied:
Agreed on the boolean reversal. I don't know if we can fit it but to me this is best described as Fallback To Best Past Release. Some version of that may help a little in the variable/param naming as well. (fallbackToBestPastRelease)

neisw replied:
I don't really disagree and likely wouldn't have coded it this way without prior art but what I did is consistent with what is [already there](https://github.com/openshift/sippy/blob/4fd58999afbe0caddd520af91f2b17d0dd029db2/sippy-ng/src/component_readiness/AdvancedOptions.js#L24) so I'm a little hesitant to refactor.

neisw replied:
With the move to include both the fallback basis as well as the requested basis I'm leaning towards removing this as an option and letting the default config (view?) determine if we use fallback or not.

neisw replied:
I see the [views](https://github.com/openshift/sippy/blob/master/config/views.yaml#L58)  are used to control the advanced options so I will investigate that vs. removal.

neisw replied:
Reversed the boolean and changed the label / parameter naming

sosiouxme commented on sippy-ng/src/component_readiness/CompReadyTestReport.js:
```diff
@@ -504,87 +438,41 @@ View the test details report at ${document.location.href}
+            onChange={handleTabChange}
+            aria-label="Test Report Tabs"
+          >
+            <Tab label="Basis Tests" {...tabProps(0)} />
+            <Tab label="Override Tests" {...tabProps(1)} />
```
These titles are too ambiguous for a newb. It's admittedly difficult to come up with better. Maybe:
```suggestion
            <Tab label="Severest Result" {...tabProps(0)} />
            <Tab label="Specified Result" {...tabProps(1)} />
```
Again some kind of explanation for newbs would be helpful... though I'm not sure tooltips work well on tabs

neisw replied:
I haven't looked into the tooltip yet but how does this look:

![image](https://github.com/user-attachments/assets/408c530d-547b-456a-a933-fda48f69b4e2)

xueqzhan replied:
I like this one better. I was confused with the terms as well. But this makes it clear.

sosiouxme reviewed (commented):
rebase to include https://github.com/openshift/sippy/pull/2029

stbenjam commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -608,6 +677,180 @@ func (s *sampleQueryGenerator) queryTestStatus() (crtype.ReportTestStatus, []err
+		}
+
+		wg.Add(1)
+		go func(queryRelease *crtype.Release, queryStart, queryEnd time.Time) {
```
This isn't the only case where we're not doing it, but we need to start propagating the context from the http Request down into component readiness and using it.  When a client disconnects, these still continue executing which isn't ideal.

At the minimum it makes sippy a target for a ddos attack (clients can quickly reconnect over and over again and spawn off many goroutines that stay running).

i.e.:

```golang
func (s *Server) jsonComponentReportFromBigQuery(w http.ResponseWriter, req *http.Request) {
    ctx := req.Context()

    // [....]
    componentreadiness.GetComponentReportFromBigQuery(ctx, ...)
    // [...]
}

func GetComponentReportFromBigQuery(ctx context.Context, ...) {
    // [...]
    go func() {
        select {
        case <-ctx.Done():
            return // terminate when context does
        default:
            // actual body of work to do
        }
    }()
    // [...]
}

```

stbenjam replied:
I filed a separate card for this: https://issues.redhat.com/browse/TRT-1845

stbenjam reviewed (commented):
- Can you share a localhost link to a page with a fallback? It'll make it easier if I'm running locally 

- Are you seeing any errors with redis locally?  About 25% of the time things fail to cache - maybe it's my local network, the redis is running elsewhere

```
INFO[2024-10-03T07:28:16.163-04:00] Fallback (4.15) QueryTestStatus completed in 23.094652875s with 180541 base results from db
WARN[2024-10-03T07:28:16.797-04:00] couldn't persist new item to cache            error="write tcp 192.168.1.157:62924->192.168.1.215:6379: i/o timeout"
```

- Is there any way to wire up ignoreFallback to the cache being unavailable? Running locally if you don't have a cache, it's very very slow


Some UI comments:
- I think we should focus on only showing the user one square / one assessment per test.  Having 2 muddies the waters: which is the right one? You can put it in a tooltip that it was overridden and what the original result was

- Would it be possible to show all the releases in tabs? e.g. the tab titles could be the basis release, and let me see all the ones we have (e.g. all 4).  It'd help for archeology, i.e. to see how the test did over time

stbenjam commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -608,6 +677,180 @@ func (s *sampleQueryGenerator) queryTestStatus() (crtype.ReportTestStatus, []err
+		end := f.BaseEnd
+		var crRelease *crtype.Release
+
+		for _, release := range releases {
+			if release.Release == fallbackRelease {
+				crRelease = &release
```
```suggestion
		for i, release := range releases {
			if release.Release == fallbackRelease {
				crRelease = &releases[i]
```

Or [update go.mod to require 1.22](https://go.dev/blog/loopvar-preview)

neisw commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -608,6 +677,180 @@ func (s *sampleQueryGenerator) queryTestStatus() (crtype.ReportTestStatus, []err
+
+		}(crRelease, start, end)
+
+		// this gets removed when we stop loading fallback data for the current base release
+		// and the section up to gets uncommented
+		fallbackRelease, err = previousRelease(fallbackRelease)
+		if err != nil {
+			log.WithError(err).Errorf("Failure determining fallback release for %s", fallbackRelease)
+			continue
+		}
```
Nice catch, thank you!

neisw commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -1321,6 +1579,96 @@ func (c *componentReportGenerator) getRequiredConfidence(testID string, variants
+				}
+			} else {
+				// if we miss on a prior release we stop
+				break
```
In the case of previousReleae err would be != nil and the for loop would end but I can make it more explicit.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -608,6 +677,180 @@ func (s *sampleQueryGenerator) queryTestStatus() (crtype.ReportTestStatus, []err
+				if release.GADate != nil && release.GA30DayPrior != nil {
+					end = *release.GADate
+					start = *release.GA30DayPrior
+				}
```
So unless if the fallback feature is enabled, regardless what base date range you choose, we may swap it out to the 30 days prior to the selected release's GA?

I'm wondering how we can smooth over the UX there. We've got the relative strings somewhat available via the views work (now-7d, ga-30d), but users can't specify that in the UI today. Seems like we *must* preserve ability to select specific date ranges, we can't just assume to let someone type/select now/ga-Xd because someone will eventually want a specific past date that is neither. Personally I would not be opposed to just typing in ga-30d in a textfield and adding support for 2024-10-14-30d (or some less dash ridden format) Frankly it's quicker than a date picker IMO.

stbenjam replied:
<img width="225" alt="image" src="https://github.com/user-attachments/assets/c9c31e08-0b38-4927-b2da-0509f96320b7">

I wish the basis release was a multi-select rather than having a fallback toggle of some kind.

For the dates, I'm not that convinced a text box is a great UI for someone who hasn't used CR before.  You could just disable the date picker when multi-release is selected and force GA. Date picker is more natural. Jira does dropdown filters by default, and has an advanced jql input field, which could also be an option.

BTW, we need to fix 30 days vs 4 weeks (28) and pick just one

neisw replied:
We only override dates for fallback queries.  We don't change the date for the selected basis.

We might consider something like `default`, or allow users to select specific releases / with dates.

I think much of this is likely separate from the initial effort though.  Lots we can do to standardize date selection / ui / etc.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -415,45 +453,49 @@ func (c *componentReportGenerator) getCommonTestStatusQuery(allJobVariants crtyp
-	}
+	// fallback queries get all variants with no filtering
+	// so all tests are fetched then cached
+	if !isFallback {
```
Is there a in-code replacement for the removed IncludeVariants query filtering? Presumably if we skip it here we must do it in code later, but I couldn't find it in the PR. (searching for uses of IncludeVariants)

And followup question, can this function just change to always NOT filter on variants/capability here, all the time, regardless if it's fallback. Wondering if we can avoid a logic fork and have just one path through here, and always rely on our own code for the filtering.

neisw replied:
Capturing the discussion we had on this, swapping out the basis for the sample in the loop leaves us unable to detect missing samples.  Will need some work there to report on tests that existed in the base but not the sample when we look to refactor this.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -350,7 +388,7 @@ func (c *componentReportGenerator) GenerateComponentReportTestStatus() (crtype.R
 
 // getCommonTestStatusQuery returns the common query for the higher level summary component summary.
-func (c *componentReportGenerator) getCommonTestStatusQuery(allJobVariants crtype.JobVariants, isSample bool) (string, string, []bigquery.QueryParameter) {
+func (c *componentReportGenerator) getCommonTestStatusQuery(allJobVariants crtype.JobVariants, isSample, isFallback bool) (string, string, []bigquery.QueryParameter) {
```
Our production redis pod looks to run pretty consistently around 500mb.

Locally with a little testing I'm around 2.2gb. I believe this is with all main views cached so that number isn't horrible, but it looks like you will need to bump the redis memory request up from 1gb to maybe 5-10, otherwise we'll be cache missing too often as things may get evicted.

neisw replied:
Let's discuss 135a2c0711cf31b74918a9267f6e83f1a13c3992 and possibly a separate redis deployment for these large caches so we don't impact the main report cache.  Having redis in the persistent cache layer is optional but does speed things up.  So nice to have but if something is evicted we can still get it quicker from the persistent backend than completely redoing the query.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -1321,6 +1579,96 @@ func (c *componentReportGenerator) getRequiredConfidence(testID string, variants
 	return c.RequestAdvancedOptions.Confidence
 }
 
+func (c *componentReportGenerator) matchBaseRegression(testID crtype.ReportTestIdentification, baseRelease string, baseStats crtype.TestStatus) (crtype.TestStatus, string) {
```
Taking a pass at godoc would be helpful on some of these more important functions.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -1347,20 +1706,46 @@ func (c *componentReportGenerator) generateComponentTestReport(baseStatus map[st
+			baseStats, baseRelease, testStats = c.matchBestBaseStats(testID, testIdentification, baseRelease, baseStats, sampleStats, requiredConfidence, approvedRegression, resolvedIssueCompensation)
+
+			if baseRelease != c.BaseRelease.Release {
+				log.Infof("Overrode base stats using release %s for Test: %s - %s", baseRelease, baseStats.TestName, testIdentification)
```
There's now the explanations field on a regressed test that could be used to show this info in the tooltip and test details page if desired.

neisw replied:
Done, thanks!

neisw commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -1347,20 +1706,46 @@ func (c *componentReportGenerator) generateComponentTestReport(baseStatus map[st
+			}
+
+			// this is where we look to see if a previous release has a higher pass rate
+			baseRelease := c.BaseRelease.Release
```
Hmm, but it may not actually be the requestedBaseRelease, the next line down looks to return a fallback if that is determined higher threshold.  Would 'matchedBaseRelease' work / be clearer?

neisw commented:
> * Can you share a localhost link to a page with a fallback? It'll make it easier if I'm running locally
> 
>     * Are you seeing any errors with redis locally?  About 25% of the time things fail to cache - maybe it's my local network, the redis is running elsewhere
> 
> 
> ```
> INFO[2024-10-03T07:28:16.163-04:00] Fallback (4.15) QueryTestStatus completed in 23.094652875s with 180541 base results from db
> WARN[2024-10-03T07:28:16.797-04:00] couldn't persist new item to cache            error="write tcp 192.168.1.157:62924->192.168.1.215:6379: i/o timeout"
> ```
> 
>     * Is there any way to wire up ignoreFallback to the cache being unavailable? Running locally if you don't have a cache, it's very very slow
> 
> 
> Some UI comments:
> 
>     * I think we should focus on only showing the user one square / one assessment per test.  Having 2 muddies the waters: which is the right one? You can put it in a tooltip that it was overridden and what the original result was
> 
>     * Would it be possible to show all the releases in tabs? e.g. the tab titles could be the basis release, and let me see all the ones we have (e.g. all 4).  It'd help for archeology, i.e. to see how the test did over time

Links for [Component Readiness](http://localhost:3000/sippy-ng/component_readiness/main?baseEndTime=2024-08-30%2023%3A59%3A59&baseRelease=4.17&baseStartTime=2024-08-02%2000%3A00%3A00&columnGroupBy=Platform%2CArchitecture%2CNetwork&confidence=95&ignoreDisruption=1&ignoreFallbackBasis=0&ignoreMissing=0&includeVariant=Architecture%3Aamd64&includeVariant=FeatureSet%3Adefault&includeVariant=Installer%3Aipi&includeVariant=Installer%3Aupi&includeVariant=Owner%3Aeng&includeVariant=Platform%3Aaws&includeVariant=Platform%3Aazure&includeVariant=Platform%3Agcp&includeVariant=Platform%3Ametal&includeVariant=Platform%3Avsphere&includeVariant=Topology%3Aha&includeVariant=CGroupMode%3Av2&includeVariant=ContainerRuntime%3Arunc&minFail=3&pity=5&regressedModal=0&sampleEndTime=2024-09-28%2023%3A59%3A59&samplePRNumber=&samplePROrg=&samplePRRepo=&sampleRelease=4.18&sampleStartTime=2024-09-21%2000%3A00%3A00&view=) and [Test Details](http://localhost:3000/sippy-ng/component_readiness/test_details?Aggregation=none&Architecture=amd64&Architecture=amd64&FeatureSet=default&FeatureSet=default&Installer=ipi&Installer=ipi&Network=ovn&Network=ovn&NetworkAccess=default&Platform=azure&Platform=azure&Scheduler=default&SecurityMode=default&Suite=unknown&Suite=unknown&Topology=ha&Topology=ha&Upgrade=none&Upgrade=none&baseEndTime=2024-08-30%2023%3A59%3A59&baseRelease=4.17&baseStartTime=2024-08-02%2000%3A00%3A00&capability=operator-conditions&columnGroupBy=Platform%2CArchitecture%2CNetwork&component=Etcd&confidence=95&dbGroupBy=Platform%2CArchitecture%2CNetwork%2CTopology%2CFeatureSet%2CUpgrade%2CSuite%2CInstaller&environment=amd64%20default%20ipi%20ovn%20azure%20unknown%20ha%20none&ignoreDisruption=true&ignoreFallbackBasis=false&ignoreMissing=false&includeVariant=Architecture%3Aamd64&includeVariant=CGroupMode%3Av2&includeVariant=ContainerRuntime%3Arunc&includeVariant=FeatureSet%3Adefault&includeVariant=Installer%3Aipi&includeVariant=Installer%3Aupi&includeVariant=Owner%3Aeng&includeVariant=Platform%3Aaws&includeVariant=Platform%3Aazure&includeVariant=Platform%3Agcp&includeVariant=Platform%3Ametal&includeVariant=Platform%3Avsphere&includeVariant=Topology%3Aha&minFail=3&pity=5&sampleEndTime=2024-09-28%2023%3A59%3A59&sampleRelease=4.18&sampleStartTime=2024-09-21%2000%3A00%3A00&testBasisRelease=4.15&testId=Operator%20results%3A45d55df296fbbfa7144600dce70c1182&testName=operator%20conditions%20etcd) should work for you.  It is the Etcd regression for operator condition that is hitting fallback.  I haven't noticed issues persisting to the cache but will keep an eye out for it.  I removed the extra status icon and changed the tabs to reflect the version of the Basis they are for.  I could see a uses case for showing them all (like in your screenshot for selecting multiple releases) we can discuss more but at the moment I'm thinking it would be new work and not part of the first effort.

neisw commented:
135a2c0711cf31b74918a9267f6e83f1a13c3992 introduces both a compression and persistent layer for caching.

There are currently some redundant config options for dataset and project that I anticipate will go away if/when we formalize this by adding the cache dataset to our main project.  

Additionally it takes a separate 'persistent-warm-cache' url to allow a separate cache from the main cache, if desired, to reduce evicting items from the main cache.  

There is also the option to be 'read-only'  for the persistent cache in the case of main sippy being read only.  We would need to coordinate another process like regression tracking to populate the persistent cache data.

```
--redis-url=redis://localhost:6379 --persistent-cache-dataset=fsbabcock_test --persistent-cache-project=openshift-ci-data-analysis --persistent-warm-cache-url=redis://localhost:6380 --persistent-cache-duration=12h --persistent-cache-write=true 
```

dgoodwin commented on pkg/cache/compressed/cache.go:
```diff
@@ -0,0 +1,106 @@
+	Cache cache.Cache
+}
+
+func NewCompressedCache(c cache.Cache) (*Cache, error) {
```
Doesn't seem to be used at this time? Assume this was an attempt at helping us stick with just redis.

neisw replied:
Updated to actually use the function and not just init the struct.

dgoodwin commented on pkg/cache/bigquerycache/bigquery.go:
```diff
@@ -0,0 +1,223 @@
+	chunkSize       = 7000000 // ~7MB to stay under the max row limit
+)
+
+type Cache struct {
```
Godoc would be good here, what this is for, why we did it, how it interacts with the warm redis cache.

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -504,6 +545,40 @@ func (c *componentReportGenerator) getBaseQueryStatus(ctx context.Context,
+	fcache := c.client.Cache
+	if c.client.PersistentCache != nil {
+		fcache = c.client.PersistentCache
+	}
```
This feels a little off in ways that will be confusing to explain to someone. The passing of a cache, but there's another cache embedded on the bigquery client that can be used for other purposes but also proxy off the original cache. Could we come up with a more cohesive mechanism for this module?

Passing in the bigquery cache separately would be one step clearer in my mind, its unexpected it would be hanging off the client as the two then depend on each other.

But ideally a single cache that could handle all the logic would be really nice. It could know to make use or the bigquery cache or not based on the size of the data and the duration requested to cache for. Though I don't know how we could tell it not to fallback to the bigquery cache for most regular pre-existing keys.

neisw replied:
If there is support for the change we could move everything to a single cache that would only attempt to persist items that have a retention time longer than X and/or byte length > Y, etc.

dgoodwin replied:
That would be excellent and really nicely encapsulate the complexity.

neisw replied:
Had to modify the Get url to include duration but added logic in the bigquery cache to not persist data below a min duration and collapsed down to a single cache from the component_report perspective.

dgoodwin commented on pkg/cache/bigquerycache/bigquery.go:
```diff
@@ -0,0 +1,223 @@
+	ChunkIndex int       `bigquery:"chunk_index"`
+}
+
+func (c Cache) Get(ctx context.Context, key string) ([]byte, error) {
```
It would be very helpful here to instrument this and Set. Prometheus counters for cache hits and missed on warm vs cold. Histograms for the size of the keys being set if you could predict ~10 bucket sizes. I'd settle for logging though. 

Being able to see how it's behaving would be quite beneficial.

neisw replied:
Added some for hits, misses and sets.  Data sizes are in the logs now but we can pursue histograms if there is value.
Sample values from deleting redis pod and also forcing refresh
```
# HELP sippy_persistent_backend_cache_get Number of persistent cache get that query the backend cache.
# TYPE sippy_persistent_backend_cache_get counter
sippy_persistent_backend_cache_get 1
# HELP sippy_persistent_cache_backend_set Number of persistent cache backend set calls.
# TYPE sippy_persistent_cache_backend_set counter
sippy_persistent_cache_backend_set 5
# HELP sippy_persistent_cache_get Number of persistent cache get requests.
# TYPE sippy_persistent_cache_get counter
sippy_persistent_cache_get 47
# HELP sippy_persistent_warm_cache_miss Number of persistent cache gets that were not in the warm cache.
# TYPE sippy_persistent_warm_cache_miss counter
sippy_persistent_warm_cache_miss 15
```

dgoodwin reviewed (commented):
Looking great. My main concern is the increase in complexity so most of these comments are related to trying to keep things as clear as is reasonable.

neisw commented on pkg/flags/cache.go:
```diff
@@ -23,6 +34,43 @@ func (f *CacheFlags) BindFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&f.PersistentWarmCacheURL,
+		"persistent-warm-cache-url",
+		os.Getenv("PERSISTENT_WARM_CACHE_URL"),
+		"Redis URL for persistent warm caching")
```
Yep, or I can drop this config for now.  I wanted to allow for separate instances but if we don't want it currently we can simplify and always add it back later.

neisw replied:
Done

neisw commented on pkg/cache/compressed/cache_test.go:
```diff
@@ -0,0 +1,74 @@
+	cache, err := NewCompressedCache(&PseudoCache{cache: make(map[string][]byte)})
+
+	if err != nil {
+		t.Fatalf("Failed to create compression cache")
```
Thanks!

neisw commented on pkg/cache/compressed/cache.go:
```diff
@@ -0,0 +1,106 @@
+
+func (c Cache) Set(ctx context.Context, key string, content []byte, duration time.Duration) error {
+	// append the checksum
+	data, checksum, err := compress(content)
```
Should see the following entries now:
```
"Compressed cache start size: 468901.25k, after compression: 12475.21k.  Ratio: 99.97%"
```

dgoodwin replied:
Very cool!

dgoodwin commented:
Great changes, thanks.

This looks good to try to me.

/lgtm
/hold

neisw commented on pkg/bigquery/client.go:
```diff
@@ -11,9 +11,10 @@ import (
-	Dataset string
+	BQ              *bigquery.Client
+	Cache           cache.Cache
+	PersistentCache cache.Cache
```
Simplified to single cache and doc'd the bigquery cache

neisw commented on pkg/cache/bigquerycache/bigquery.go:
```diff
@@ -0,0 +1,302 @@
+// it will not check for an entry but instead return a miss.
+type Cache struct {
+	client        *sippybq.Client
+	readOnly      bool
```
Wondering if we need an inverse of this that is `primaryWrite` indicating skip the warm cache lookup and go to the backend cache.  In the case above where maxExpiration is shortened for the cache write process that won't impact the read from the warm cache which would lead to it not considering the data expired and therefore not refreshing the cache as expected.

neisw commented on pkg/cache/bigquerycache/bigquery.go:
```diff
@@ -0,0 +1,302 @@
+	persistentCacheGetMetric.WithLabelValues().Inc()
+
+	// if we have it in our warm cache use it
+	if c.client.Cache != nil {
```
&& (duration < c.minExpiration || !c.primaryWrite)

dgoodwin commented:
/lgtm

Love the metrics, thanks, they may actually prove useful in questions like, was someone hitting comp readiness hard over the weekend.
